File: lib/tm11/tm11.cpp
#include <Arduino.h>

#include <pdp11.h>
#include <unibus.h>
#include "tm11.h"
#include "cpu.h"

#include "SdFat.h"

#define DEBUG_TM11 0

#define TM_GO    01
#define TM_RCOM  02
#define TM_WCOM  04
#define TM_WEOF  06
#define TM_SFORW 010
#define TM_SREV  012
#define TM_WIRG  014
#define TM_REW   016

#define TM_CRDY  0200
#define TM_IE    0100
#define TM_CE    0100000

#define TM_TUR   01
#define TM_WRL   04
#define TM_BOT   040
#define TM_SELR  0100
#define TM_NXM   0200
#define TM_EOT   02000
#define TM_EOF   040000
#define TM_ILC   0100000

#define TAPE_LEN 6195200

#define TAPE_EOF 0x00000000
#define TAPE_EOT 0xFFFFFFFF

namespace tm11 {

    uint16_t MTS;   // 772520 Status Register	
    uint16_t MTC;   // 772522 Command Register		
    uint16_t MTBRC; // 772524 Byte Record Counter
    uint16_t MTCMA; // 772526 Current Memory Address Register
    uint16_t MTD;   // 772530 Data Buffer
    uint16_t MTRD;  // 772532 TU10 Read Lines

    uint16_t rewind;

    uint16_t sector, address, count;

    struct tape tmdata[TM_NUM_DRV];

    void reset() {
        MTS = TM_TUR;
        MTC = TM_CRDY;
        for (int i = 0; i < 8; i++) {
            tmdata[i].pos = 0;
        }
        MTBRC = 0;
        MTCMA = 0;        
        MTRD = 0;
        MTD = 0;
    }

    uint16_t read16(uint32_t a) {
        /*
        if (DEBUG_TM11) {
            Serial.printf("tm11: read16 : %06o\r\n", a);
        }
        */
        switch(a & ~1) {
            case 0772520:
                return MTS;
            case 0772522:
                MTC &= 0x7FFF;
                if (MTS & 0xFF80) MTC |= 0x8000;
                return MTC;
            case 0772524:
                return MTBRC;
            case 0772526:
                return MTCMA;   
            case 0772530:
                return MTD;
            case 0772532:
                return MTRD; 
            default:
                Serial.printf("tm11: invalid read16 %06o\r\n", a);
                return 0;
        }                
    }

    void write16(uint32_t a, uint16_t v) {     
        
        if (DEBUG_TM11) {
            Serial.printf("tm11: write16: %06o: %06o\r\n", a, v);
        }
        
        switch(a) {
            case 0772522:
                MTC &= (TM_CRDY|TM_CE);
                MTC |= v & ~(TM_CRDY|TM_CE);
                if (MTC & (TM_GO|TM_IE)) {
                    MTC &= ~TM_CRDY;
                    go();
                }
                break;
            case 0772524:
                MTBRC = v;
                break;
            case 0772526:
                MTCMA = v;  
                break;
            case 0772530:
                MTD = v;
                break;
            case 0772532:
                break;
            default:
                Serial.printf("tm11: invalid write16 %06o: %06o\r\n", a, v);
        }
    }

    uint16_t cmd_end() {
        if (rewind) {
           rewind--;
           if (!rewind) {                
                MTS &= ~2;   // clear rewind
                MTS |= 0x21; // set bot and tur
                MTC |= 0x80; // set cur
                Serial.printf("tm11: rewind done, MTS: %06o, MTC: %06o\r\n", MTS, MTC);
           }
        }

        MTS |= 1;
        MTC |= 0x80;
        return MTC & 0x40;
    }

    void finish() {
        if (MTS & (TM_ILC|TM_EOT|TM_NXM))
            MTC |= TM_CE;
        MTC |= TM_CRDY;
        if (MTC & TM_IE) {
            MTC &= ~TM_IE;
            cpu::interrupt(INTTM, 5);
        }
        if (DEBUG_TM11) {
            Serial.printf("tm11: MTS: %06o, MTC: %06o, MTBRC: %06o, MTCMA: %06o, fin\r\n", MTS, MTC, MTBRC, MTCMA);
        }
        cmd_end();
    }

    void go() {
        uint8_t drive = (MTC >> 8) & 3;
        bool attached = tmdata[drive].attached;
        off_t pos = tmdata[drive].pos;
        if (!attached) {
            Serial.printf("tm11: drive %d is not attached\r\n", drive);
            longjmp(trapbuf, INTBUS);
            return;
        }

        MTC &= ~TM_CE;
        MTS &= ~(TM_ILC|TM_NXM);
        
        uint8_t cmd = (MTC >> 1) & 7;
        if (DEBUG_TM11) {
            Serial.printf("tm11: MTS: %06o, MTC: %06o, MTBRC: %06o, MTCMA: %06o, cmd: %d, pos: %d\r\n", MTS, MTC, MTBRC, MTCMA, cmd, pos);
        }
        switch (cmd) {
            case 0: { // off-line
                // clr tur? and cur?
                break;
            }
            case 1: { // read           
                MTS &= ~TM_EOF;
                uint32_t addr = (MTC & 060) << 12 | MTCMA;
                
                if (DEBUG_TM11) {
                    Serial.printf("tm11: read: pos %d, addr: %06o, mtbrc: %d/%06o\r\n", tmdata[drive].pos, addr, MTBRC, MTBRC);
                }
                if (tmdata[drive].pos > (off_t) tmdata[drive].file.fileSize()) {
                    MTS |= TM_EOT;
                    break;
                }
                __disable_irq();
                tmdata[drive].file.seekSet(tmdata[drive].pos);
                while(MTBRC) {
                    MTS &= ~TM_BOT;
                    uint16_t dv = tmdata[drive].file.read(); 
                    MTBRC++;
                    dv |= (tmdata[drive].file.read() << 8);    
                    MTBRC++;
                    
                    if (DEBUG_TM11) {
                        Serial.printf("tm11: read %06o -> %08o\r\n", dv, addr);
                        yield();
                    }
                    
                    unibus::write16(addr, dv);
                    addr+=2;
                    MTCMA = addr & 0xFFFF;
                    MTC  |= ((addr & 0x300000000) >> 12);
                    tmdata[drive].pos += 2;
                    if (tmdata[drive].pos > (off_t) tmdata[drive].file.fileSize()) {
                        MTS |= TM_EOT;
                        break;
                    }
                }    
                __enable_irq();
                yield();
                break;
            }
            case 2: { // write
                MTC &= ~TM_EOF;
                uint32_t addr = (MTC & 060) << 12 | MTCMA;
                if (DEBUG_TM11) {
                    Serial.printf("tm11: write: pos %d, addr: %06o, mtbrc: %d/%06o\r\n", tmdata[drive].pos, addr, MTBRC, MTBRC);
                }
                if (tmdata[drive].pos >= TAPE_LEN) {
                    MTS |= TM_EOT;
                    break;
                }
                __disable_irq();
                tmdata[drive].file.seekSet(tmdata[drive].pos);
                while(MTBRC) {
                    MTS &= ~TM_BOT;
                    if (DEBUG_TM11) {
                        Serial.printf("tm11: read %08o\r\n", addr);
                        yield();
                    }
                    uint16_t dv = unibus::read16(addr);
                    tmdata[drive].file.write(dv & 0xFF);
                    MTBRC++;
                    tmdata[drive].file.write((dv >> 8) & 0xFF);
                    MTBRC++;
                    addr+=2;
                    MTCMA = addr & 0xFFFF;
                    MTC |= ((addr & 0x300000000) >> 12);
                    tmdata[drive].pos += 2;
                }
                MTC |= TM_EOF;
                __enable_irq()
                yield();
                break;
            }
            case 3: { // write eof 
                MTC &= ~TM_EOF;
                break;
            }
            case 4: { // space forward
                Serial.println("tm11: space forward uninplemented");
                break;
            }
            case 5: { // space reverse
                Serial.println("tm11: space reverse unimplemented");
                break;
            }
            case 6: { // write with extended IRG
                Serial.println("tm11: write with extended irg uninplemented");
                break;
            }
            case 7: { // rewind
                Serial.println("tm11: rewind");                
                MTS &= ~(TM_EOT|TM_EOF);
                if (MTC & 0x40) {
                    cpu::interrupt(INTTM, 5);
                }                
                if (!tmdata[drive].file.seekSet(0)) {
                    Serial.printf("tm11: failed to seek: drive: %d, pos: %d\r\n", drive, pos);
                }                
                tmdata[drive].file.flush();
                tmdata[drive].pos = 0;
                rewind = 1;                
                break;
            }
            default: {
                Serial.printf("tm11: cmd %d uninplemented\r\n", cmd);
                if (MTC & TM_GO) {
                    MTS |= TM_ILC;
                }
                break;
            }
        }
        finish();
    }        
};

File: lib/tm11/tm11.js
// =========== TM11 routines ===========


var tm11 = {
    mts: 0x65, // 17772520 Status Register    6 selr 5 bot 2 wrl 0 tur
    mtc: 0x6080, // 17772522 Command Register   14-13 bpi 7 cu rdy
    mtbrc: 0, // 17772524 Byte Record Counter
    mtcma: 0, // 17772526 Current Memory Address Register
    mtd: 0, // 17772530 Data Buffer Register
    mtrd: 0, // 17772532 TU10 Read Lines
    meta: [] //meta data for drive
};

function tm11_commandEnd() {
    tm11.mts |= 1; // tape unit ready
    tm11.mtc |= 0x80;
    return tm11.mtc & 0x40;
}

function tm11_finish() {
    if (tm11.mtc & 0x40) {
        interrupt(0, 10, 5 << 5, 0224, tm11_commandEnd);
    } else { // if interrupt not enabled just mark completed
        tm11_commandEnd();
    }
}

function tm11_end(err, meta, position, address, count) {
    if (err == 0 && meta.command > 0) {
        if (address == 0 || address > 0x80000000) { // tape mark
            meta.position = position;
            tm11.mts |= 0x4000; // set EOF bit
        } else {
            switch (meta.command) {
                case 1: // read
                    meta.position = position + 2 + ((address + 1) >> 1);
                    meta.command = 0;
                    count = (0x10000 - tm11.mtbrc) & 0xffff;
                    if (count >= address || count == 0) {
                        count = address;
                        tm11.mtbrc = (tm11.mtbrc + count) & 0xffff;
                    } else {
                        tm11.mts |= 0x200; // RLE
                        tm11.mtbrc = 0;
                    }
                    address = ((tm11.mtc & 0x30) << 12) | tm11.mtcma;
                    diskIO(2, meta, position, address, count);
                    // calculate meta.position set count to reduced amount
                    return;
                case 4: // space forward
                    position = position + 2 + ((address + 1) >> 1);
                    meta.position = position;
                    tm11.mtbrc = (tm11.mtbrc + 1) & 0xffff;
                    if (tm11.mtbrc) {
                        diskIO(4, meta, position, 0, 4);
                        return;
                    }
                    break;
                case 5: // space reverse
                    position = position - 4 - ((address + 1) >> 1);
                    meta.position = position;
                    tm11.mtbrc = (tm11.mtbrc + 1) & 0xffff;
                    if (tm11.mtbrc) {
                        if (position > 0) {
                            diskIO(4, meta, position - 2, 0, 4);
                            return;
                        }
                    }
                    break;
                default:
                    panic();
            }
        }
    }
    if (meta.command == 0) {
        tm11.mtbrc = (tm11.mtbrc - count) & 0xffff;
        tm11.mtcma = address & 0xffff;
        tm11.mtc = (tm11.mtc & ~0x30) | ((address >> 12) & 0x30);
    }
    switch (err) {
        case 1: // read error
            tm11.mts |= 0x100; // Bad tape error
            break;
        case 2: // NXM
            tm11.mts |= 0x80; // NXM
            break;
    }
    tm11_finish();
}

function tm11_init() {
    var i;
    tm11.mts = 0x65; //  6 selr 5 bot 2 wrl 0 tur
    tm11.mtc = 0x6080; //  14-13 bpi 7 cu rdy
    for (i = 0; i < 8; i++) {
        if (typeof tm11.meta[i] !== "undefined") {
            tm11.meta[i].position == 0;
        }
    }
}

function tm11_go() {
    var sector, address, count;
    var drive = (tm11.mtc >> 8) & 3;
    tm11.mtc &= ~0x81; // ready bit (7!) and go (0)
    tm11.mts &= 0x04fe; // turn off tape unit ready
    if (typeof tm11.meta[drive] === "undefined") {
        tm11.meta[drive] = {
            "cache": [],
            "blockSize": 65536,
            "postProcess": tm11_end,
            "drive": drive,
            "mapped": 1,
            "maxblock": 0,
            "position": 0,
            "command": 0,
            "url": "tm" + drive + ".tap"
        };
    }
    tm11.meta[drive].command = (tm11.mtc >> 1) & 7;
    //console.log("TM11 Function "+(tm11.meta[drive].command).toString(8)+" "+tm11.mtc.toString(8)+" "+tm11.mts.toString(8)+" @ "+tm11.meta[drive].position.toString(8));
    switch (tm11.meta[drive].command) { // function code
        case 0: // off-line
            break;
        case 1: // read
            diskIO(4, tm11.meta[drive], tm11.meta[drive].position, 0, 4);
            return;
        case 2: // write
        case 3: // write end of file
        case 6: // write with extended IRG
            break;
        case 4: // space forward
            diskIO(4, tm11.meta[drive], tm11.meta[drive].position, 0, 4);
            return;
        case 5: // space reverse
            if (tm11.meta[drive].position > 0) {
                diskIO(4, tm11.meta[drive], tm11.meta[drive].position - 2, 0, 4);
                return;
            }
            break;
        case 7: // rewind
            tm11.meta[drive].position = 0;
            tm11.mts |= 0x20; // set BOT
            break;
        default:
            break;
    }
    tm11_finish();
}

function accessTM11(physicalAddress, data, byteFlag) {
    var result, drive = (tm11.mtc >> 8) & 3;
    switch (physicalAddress & ~1) {
        case 017772520: // tm11.mts
            tm11.mts &= ~0x20; // turn off BOT
            if (typeof tm11.meta[(tm11.mtc >> 8) & 3] !== "undefined") {
                if (tm11.meta[(tm11.mtc >> 8) & 3].position == 0) {
                    tm11.mts |= 0x20; // turn on BOT
                }
            }
            result = tm11.mts;
            break;
        case 017772522: // tm11.mtc
            tm11.mtc &= 0x7fff; // no err bit
            if (tm11.mts & 0xff80) tm11.mtc |= 0x8000;
            result = insertData(tm11.mtc, physicalAddress, data, byteFlag);
            if (data >= 0 && result >= 0) {
                if ((tm11.mtc & 0x40) && !(result & 0x40)) { // if IE being reset then kill any pending interrupts
                    interrupt(1, -1, 5 << 5, 0224);
                }
                if (result & 0x1000) { //init
                    tm11.mts = 0x65; //  6 selr 5 bot 2 wrl 0 tur
                    tm11.mtc = 0x6080; //  14-13 bpi 7 cu rdy
                }
                if ((tm11.mtc & 0x80) && (result & 0x1)) {
                    tm11.mtc = (tm11.mtc & 0x80) | (result & 0xff7f);
                    tm11_go();
                } else {
                    if ((result & 0x40) && (tm11.mtc & 0xc0) == 0x80) {
                        interrupt(1, 10, 5 << 5, 0224);
                    }
                    tm11.mtc = (tm11.mtc & 0x80) | (result & 0xff7f);
                }
            }
            break;
        case 017772524: // tm11.mtbrc
            result = insertData(tm11.mtbrc, physicalAddress, data, byteFlag);
            if (result >= 0) tm11.mtbrc = result;
            break;
        case 017772526: // tm11.mtcma
            result = insertData(tm11.mtcma, physicalAddress, data, byteFlag);
            if (result >= 0) tm11.mtcma = result;
            break;
        case 017772530: // tm11.mtd
        case 017772532: // tm11.mtrd
            result = 0;
            break;
        default:
            CPU.CPU_Error |= 0x10;
            return trap(4, 134);
    }
    //console.log("TM11 Access "+physicalAddress.toString(8)+" "+data.toString(8)+" "+byteFlag.toString(8)+" -> "+result.toString(8));
    return result;
}
